File: clients/go/consensus/sighash.go
package consensus

type SighashV1PrehashCache struct {
	tx                 *Tx
	hashOfDaCoreFields [32]byte
	hashAllPrevouts    [32]byte
	hashAllSequences   [32]byte
	hashAllOutputs     [32]byte
	singleOutputs      map[uint32][32]byte
}

func IsValidSighashType(sighashType uint8) bool {
	switch sighashType {
	case SIGHASH_ALL, SIGHASH_NONE, SIGHASH_SINGLE,
		SIGHASH_ALL | SIGHASH_ANYONECANPAY,
		SIGHASH_NONE | SIGHASH_ANYONECANPAY,
		SIGHASH_SINGLE | SIGHASH_ANYONECANPAY:
		return true
	default:
		return false
	}
}

func SighashV1Digest(tx *Tx, inputIndex uint32, inputValue uint64, chainID [32]byte) ([32]byte, error) {
	return SighashV1DigestWithType(tx, inputIndex, inputValue, chainID, SIGHASH_ALL)
}

func NewSighashV1PrehashCache(tx *Tx) (*SighashV1PrehashCache, error) {
	if tx == nil {
		return nil, txerr(TX_ERR_PARSE, "sighash: nil tx")
	}
	daCoreBytes, err := daCoreFieldsBytes(tx)
	if err != nil {
		return nil, err
	}
	prevouts := make([]byte, 0, len(tx.Inputs)*(32+4))
	sequences := make([]byte, 0, len(tx.Inputs)*4)
	for _, txIn := range tx.Inputs {
		prevouts = append(prevouts, txIn.PrevTxid[:]...)
		prevouts = AppendU32le(prevouts, txIn.PrevVout)
		sequences = AppendU32le(sequences, txIn.Sequence)
	}
	outputsBytes := make([]byte, 0, len(tx.Outputs)*64)
	for _, o := range tx.Outputs {
		outputsBytes = AppendU64le(outputsBytes, o.Value)
		outputsBytes = AppendU16le(outputsBytes, o.CovenantType)
		outputsBytes = AppendCompactSize(outputsBytes, uint64(len(o.CovenantData)))
		outputsBytes = append(outputsBytes, o.CovenantData...)
	}
	return &SighashV1PrehashCache{
		tx:                 tx,
		hashOfDaCoreFields: sha3_256(daCoreBytes),
		hashAllPrevouts:    sha3_256(prevouts),
		hashAllSequences:   sha3_256(sequences),
		hashAllOutputs:     sha3_256(outputsBytes),
		singleOutputs:      make(map[uint32][32]byte),
	}, nil
}

func SighashV1DigestWithType(tx *Tx, inputIndex uint32, inputValue uint64, chainID [32]byte, sighashType uint8) ([32]byte, error) {
	cache, err := NewSighashV1PrehashCache(tx)
	if err != nil {
		return [32]byte{}, err
	}
	return SighashV1DigestWithCache(cache, inputIndex, inputValue, chainID, sighashType)
}

func SighashV1DigestWithCache(cache *SighashV1PrehashCache, inputIndex uint32, inputValue uint64, chainID [32]byte, sighashType uint8) ([32]byte, error) {
	var zero [32]byte
	if cache == nil || cache.tx == nil {
		return zero, txerr(TX_ERR_PARSE, "sighash: nil cache")
	}
	tx := cache.tx
	if int(inputIndex) < 0 || int(inputIndex) >= len(tx.Inputs) {
		return zero, txerr(TX_ERR_PARSE, "sighash: input_index out of bounds")
	}
	if !IsValidSighashType(sighashType) {
		return zero, txerr(TX_ERR_SIGHASH_TYPE_INVALID, "sighash: invalid sighash_type")
	}

	baseType := sighashType & 0x1f
	anyoneCanPay := (sighashType & SIGHASH_ANYONECANPAY) != 0

	in := tx.Inputs[inputIndex]

	var hashPrevouts [32]byte
	var hashSequences [32]byte
	if anyoneCanPay {
		prevouts := make([]byte, 0, 32+4)
		prevouts = append(prevouts, in.PrevTxid[:]...)
		prevouts = AppendU32le(prevouts, in.PrevVout)
		hashPrevouts = sha3_256(prevouts)

		sequences := make([]byte, 0, 4)
		sequences = AppendU32le(sequences, in.Sequence)
		hashSequences = sha3_256(sequences)
	} else {
		hashPrevouts = cache.hashAllPrevouts
		hashSequences = cache.hashAllSequences
	}

	var hashOutputs [32]byte
	switch baseType {
	case SIGHASH_ALL:
		hashOutputs = cache.hashAllOutputs
	case SIGHASH_NONE:
		hashOutputs = sha3_256(nil)
	case SIGHASH_SINGLE:
		hashOutputs = cache.singleOutputHash(inputIndex)
	default:
		return zero, txerr(TX_ERR_SIGHASH_TYPE_INVALID, "sighash: invalid base_type")
	}

	preimage := make([]byte, 0, 256)
	preimage = append(preimage, []byte("RUBINv1-sighash/")...)
	preimage = append(preimage, chainID[:]...)
	preimage = AppendU32le(preimage, tx.Version)
	preimage = append(preimage, tx.TxKind)
	preimage = AppendU64le(preimage, tx.TxNonce)
	preimage = append(preimage, cache.hashOfDaCoreFields[:]...)
	preimage = append(preimage, hashPrevouts[:]...)
	preimage = append(preimage, hashSequences[:]...)
	preimage = AppendU32le(preimage, inputIndex)
	preimage = append(preimage, in.PrevTxid[:]...)
	preimage = AppendU32le(preimage, in.PrevVout)
	preimage = AppendU64le(preimage, inputValue)
	preimage = AppendU32le(preimage, in.Sequence)
	preimage = append(preimage, hashOutputs[:]...)
	preimage = AppendU32le(preimage, tx.Locktime)
	preimage = append(preimage, sighashType)

	return sha3_256(preimage), nil
}

func (c *SighashV1PrehashCache) singleOutputHash(inputIndex uint32) [32]byte {
	if c == nil || c.tx == nil {
		return sha3_256(nil)
	}
	if out, ok := c.singleOutputs[inputIndex]; ok {
		return out
	}
	var hash [32]byte
	if int(inputIndex) < len(c.tx.Outputs) {
		o := c.tx.Outputs[inputIndex]
		outputsBytes := make([]byte, 0, 64)
		outputsBytes = AppendU64le(outputsBytes, o.Value)
		outputsBytes = AppendU16le(outputsBytes, o.CovenantType)
		outputsBytes = AppendCompactSize(outputsBytes, uint64(len(o.CovenantData)))
		outputsBytes = append(outputsBytes, o.CovenantData...)
		hash = sha3_256(outputsBytes)
	} else {
		hash = sha3_256(nil)
	}
	c.singleOutputs[inputIndex] = hash
	return hash
}

File: clients/rust/crates/rubin-consensus/src/sighash.rs
use std::collections::HashMap;

use crate::compactsize::encode_compact_size;
use crate::constants::{SIGHASH_ALL, SIGHASH_ANYONECANPAY, SIGHASH_NONE, SIGHASH_SINGLE};
use crate::error::{ErrorCode, TxError};
use crate::hash::sha3_256;
use crate::tx::{da_core_fields_bytes, Tx};

pub struct SighashV1PrehashCache<'a> {
    tx: &'a Tx,
    hash_of_da_core_fields: [u8; 32],
    hash_all_prevouts: [u8; 32],
    hash_all_sequences: [u8; 32],
    hash_all_outputs: [u8; 32],
    single_outputs: HashMap<u32, [u8; 32]>,
}

pub fn is_valid_sighash_type(sighash_type: u8) -> bool {
    sighash_type == SIGHASH_ALL
        || sighash_type == SIGHASH_NONE
        || sighash_type == SIGHASH_SINGLE
        || sighash_type == (SIGHASH_ALL | SIGHASH_ANYONECANPAY)
        || sighash_type == (SIGHASH_NONE | SIGHASH_ANYONECANPAY)
        || sighash_type == (SIGHASH_SINGLE | SIGHASH_ANYONECANPAY)
}

pub fn sighash_v1_digest(
    tx: &Tx,
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
) -> Result<[u8; 32], TxError> {
    sighash_v1_digest_with_type(tx, input_index, input_value, chain_id, SIGHASH_ALL)
}

pub fn sighash_v1_digest_with_type(
    tx: &Tx,
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    sighash_type: u8,
) -> Result<[u8; 32], TxError> {
    let mut cache = SighashV1PrehashCache::new(tx)?;
    sighash_v1_digest_with_cache(&mut cache, input_index, input_value, chain_id, sighash_type)
}

impl<'a> SighashV1PrehashCache<'a> {
    pub fn new(tx: &'a Tx) -> Result<Self, TxError> {
        let hash_of_da_core_fields = sha3_256(&da_core_fields_bytes(tx)?);

        let mut prevouts = Vec::with_capacity(tx.inputs.len() * (32 + 4));
        let mut sequences = Vec::with_capacity(tx.inputs.len() * 4);
        for tx_in in &tx.inputs {
            prevouts.extend_from_slice(&tx_in.prev_txid);
            prevouts.extend_from_slice(&tx_in.prev_vout.to_le_bytes());
            sequences.extend_from_slice(&tx_in.sequence.to_le_bytes());
        }

        let mut outputs_bytes = Vec::with_capacity(tx.outputs.len() * 64);
        for o in &tx.outputs {
            outputs_bytes.extend_from_slice(&o.value.to_le_bytes());
            outputs_bytes.extend_from_slice(&o.covenant_type.to_le_bytes());
            encode_compact_size(o.covenant_data.len() as u64, &mut outputs_bytes);
            outputs_bytes.extend_from_slice(&o.covenant_data);
        }

        Ok(Self {
            tx,
            hash_of_da_core_fields,
            hash_all_prevouts: sha3_256(&prevouts),
            hash_all_sequences: sha3_256(&sequences),
            hash_all_outputs: sha3_256(&outputs_bytes),
            single_outputs: HashMap::new(),
        })
    }

    fn single_output_hash(&mut self, input_index: u32) -> [u8; 32] {
        if let Some(hash) = self.single_outputs.get(&input_index) {
            return *hash;
        }

        let idx = input_index as usize;
        let hash = if idx < self.tx.outputs.len() {
            let o = &self.tx.outputs[idx];
            let mut output_bytes = Vec::with_capacity(64);
            output_bytes.extend_from_slice(&o.value.to_le_bytes());
            output_bytes.extend_from_slice(&o.covenant_type.to_le_bytes());
            encode_compact_size(o.covenant_data.len() as u64, &mut output_bytes);
            output_bytes.extend_from_slice(&o.covenant_data);
            sha3_256(&output_bytes)
        } else {
            sha3_256(&[])
        };
        self.single_outputs.insert(input_index, hash);
        hash
    }
}

pub fn sighash_v1_digest_with_cache(
    cache: &mut SighashV1PrehashCache<'_>,
    input_index: u32,
    input_value: u64,
    chain_id: [u8; 32],
    sighash_type: u8,
) -> Result<[u8; 32], TxError> {
    let tx = cache.tx;
    let idx = input_index as usize;
    if idx >= tx.inputs.len() {
        return Err(TxError::new(
            ErrorCode::TxErrParse,
            "sighash: input_index out of bounds",
        ));
    }
    if !is_valid_sighash_type(sighash_type) {
        return Err(TxError::new(
            ErrorCode::TxErrSighashTypeInvalid,
            "sighash: invalid sighash_type",
        ));
    }

    let base_type = sighash_type & 0x1f;
    let anyone_can_pay = (sighash_type & SIGHASH_ANYONECANPAY) != 0;
    let i = &tx.inputs[idx];

    let hash_prevouts = if anyone_can_pay {
        let mut prevouts = Vec::with_capacity(32 + 4);
        prevouts.extend_from_slice(&i.prev_txid);
        prevouts.extend_from_slice(&i.prev_vout.to_le_bytes());
        sha3_256(&prevouts)
    } else {
        cache.hash_all_prevouts
    };

    let hash_sequences = if anyone_can_pay {
        let mut sequences = Vec::with_capacity(4);
        sequences.extend_from_slice(&i.sequence.to_le_bytes());
        sha3_256(&sequences)
    } else {
        cache.hash_all_sequences
    };

    let hash_outputs = match base_type {
        SIGHASH_ALL => cache.hash_all_outputs,
        SIGHASH_NONE => sha3_256(&[]),
        SIGHASH_SINGLE => cache.single_output_hash(input_index),
        _ => {
            return Err(TxError::new(
                ErrorCode::TxErrSighashTypeInvalid,
                "sighash: invalid base_type",
            ))
        }
    };

    let mut preimage = Vec::with_capacity(256);
    preimage.extend_from_slice(b"RUBINv1-sighash/");
    preimage.extend_from_slice(&chain_id);
    preimage.extend_from_slice(&tx.version.to_le_bytes());
    preimage.push(tx.tx_kind);
    preimage.extend_from_slice(&tx.tx_nonce.to_le_bytes());
    preimage.extend_from_slice(&cache.hash_of_da_core_fields);
    preimage.extend_from_slice(&hash_prevouts);
    preimage.extend_from_slice(&hash_sequences);
    preimage.extend_from_slice(&input_index.to_le_bytes());
    preimage.extend_from_slice(&i.prev_txid);
    preimage.extend_from_slice(&i.prev_vout.to_le_bytes());
    preimage.extend_from_slice(&input_value.to_le_bytes());
    preimage.extend_from_slice(&i.sequence.to_le_bytes());
    preimage.extend_from_slice(&hash_outputs);
    preimage.extend_from_slice(&tx.locktime.to_le_bytes());
    preimage.push(sighash_type);

    Ok(sha3_256(&preimage))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tx::{TxInput, TxOutput};

    fn test_tx() -> Tx {
        Tx {
            version: 1,
            tx_kind: 0x00,
            tx_nonce: 7,
            inputs: vec![
                TxInput {
                    prev_txid: [0x11; 32],
                    prev_vout: 0,
                    script_sig: vec![],
                    sequence: 1,
                },
                TxInput {
                    prev_txid: [0x22; 32],
                    prev_vout: 1,
                    script_sig: vec![],
                    sequence: 2,
                },
            ],
            outputs: vec![
                TxOutput {
                    value: 5,
                    covenant_type: 0x0000,
                    covenant_data: vec![0x01; 33],
                },
                TxOutput {
                    value: 7,
                    covenant_type: 0x0000,
                    covenant_data: vec![0x02; 33],
                },
            ],
            locktime: 0,
            witness: vec![],
            da_payload: vec![],
            da_commit_core: None,
            da_chunk_core: None,
        }
    }

    #[test]
    fn sighash_cache_matches_wrapper_digest() {
        let tx = test_tx();
        let chain_id = [0x55; 32];
        let mut cache = SighashV1PrehashCache::new(&tx).expect("cache");

        let direct =
            sighash_v1_digest_with_type(&tx, 0, 10, chain_id, SIGHASH_ALL).expect("direct");
        let cached =
            sighash_v1_digest_with_cache(&mut cache, 0, 10, chain_id, SIGHASH_ALL).expect("cached");
        assert_eq!(direct, cached);
    }

    #[test]
    fn sighash_cache_reuses_single_output_hash() {
        let tx = test_tx();
        let chain_id = [0x66; 32];
        let mut cache = SighashV1PrehashCache::new(&tx).expect("cache");

        let first = sighash_v1_digest_with_cache(&mut cache, 1, 10, chain_id, SIGHASH_SINGLE)
            .expect("first");
        assert_eq!(cache.single_outputs.len(), 1);
        let second = sighash_v1_digest_with_cache(&mut cache, 1, 10, chain_id, SIGHASH_SINGLE)
            .expect("second");
        assert_eq!(cache.single_outputs.len(), 1);
        assert_eq!(first, second);
    }
}
